plugins/GSdx/GSFastList.h

```diff
@@ -0,0 +1,162 @@
+/*
+*	Copyright (C) 2017-2017 Alessandro Vetere
+*	Copyright (C) 2007-2009 Gabest
+*	http://www.gabest.org
+*
+*  This Program is free software; you can redistribute it and/or modify
+*  it under the terms of the GNU General Public License as published by
+*  the Free Software Foundation; either version 2, or (at your option)
+*  any later version.
+*
+*  This Program is distributed in the hope that it will be useful,
+*  but WITHOUT ANY WARRANTY; without even the implied warranty of
+*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+*  GNU General Public License for more details.
+*
+*  You should have received a copy of the GNU General Public License
+*  along with GNU Make; see the file COPYING.  If not, write to
+*  the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA USA.
+*  http://www.gnu.org/copyleft/gpl.html
+*
+*/
+
+#pragma once
+
+template <class T>
+struct Element {
+	T  data;
+	uint16 next_index;
+	uint16 prev_index;
+};
+
+template <class T, uint16 INITIAL_SIZE = 4>
+class FastList {
+private:
+	// The last Element<T> of the list has next_index equal to 0
+	// m_buffer[0] is the head of the list, and m_buffer has m_size elements
+	Element<T>* m_buffer;
+	uint16 m_size;
+	uint16 m_free_indexes_stack_top;
+	// m_buffer indexes that are free to be used are stacked here [END - 1]
+	uint16* m_free_indexes_stack;
+
+public:
+	FastList() {
+		m_buffer = nullptr;
+		Clear();
+	}
+
+	~FastList() {
+		_aligned_free(m_buffer);
+	}
+
+	void Clear() {
+		// Initialize END
+		m_size = INITIAL_SIZE;
+		// Initialize arrays
+		_aligned_free(m_buffer);
+		m_buffer = (Element<T>*)_aligned_malloc(m_size * sizeof(Element<T>) + (m_size - 1) * sizeof(uint16), 64);
+		m_free_indexes_stack = (uint16*)&m_buffer[m_size];
+		// Initialize head
+		m_buffer[0].data = nullptr;
+		m_buffer[0].next_index = End();
+		m_buffer[0].prev_index = 0;
+		// Stack top is 0, bottom is END - 2
+		m_free_indexes_stack_top = 0;
+		// Index 0 reserved for head
+		for (uint16 i = 0; i < m_size - 1; i++) {
+			m_free_indexes_stack[i] = i + 1;
+		}
+	}
+
+	__forceinline uint16 First() const {
```

Comment:
A real iterator will be mucher nicer.

Here an example of the GNU STL list iterator.
```
  /**
   *  @brief A list::iterator.
   *
   *  All the functions are op overloads.
  */
  template<typename _Tp>
    struct _List_iterator
    {
      typedef _List_iterator<_Tp>                _Self;
      typedef _List_node<_Tp>                    _Node;

      typedef ptrdiff_t                          difference_type;
      typedef std::bidirectional_iterator_tag    iterator_category;
      typedef _Tp                                value_type;
      typedef _Tp*                               pointer;
      typedef _Tp&                               reference;

      _List_iterator() _GLIBCXX_NOEXCEPT
      : _M_node() { }

      explicit
      _List_iterator(__detail::_List_node_base* __x) _GLIBCXX_NOEXCEPT
      : _M_node(__x) { }

      _Self
      _M_const_cast() const _GLIBCXX_NOEXCEPT
      { return *this; }

      // Must downcast from _List_node_base to _List_node to get to value.
      reference
      operator*() const _GLIBCXX_NOEXCEPT
      { return *static_cast<_Node*>(_M_node)->_M_valptr(); }

      pointer
      operator->() const _GLIBCXX_NOEXCEPT
      { return static_cast<_Node*>(_M_node)->_M_valptr(); }

      _Self&
      operator++() _GLIBCXX_NOEXCEPT
      {
	_M_node = _M_node->_M_next;
	return *this;
      }

      _Self
      operator++(int) _GLIBCXX_NOEXCEPT
      {
	_Self __tmp = *this;
	_M_node = _M_node->_M_next;
	return __tmp;
      }

      _Self&
      operator--() _GLIBCXX_NOEXCEPT
      {
	_M_node = _M_node->_M_prev;
	return *this;
      }

      _Self
      operator--(int) _GLIBCXX_NOEXCEPT
      {
	_Self __tmp = *this;
	_M_node = _M_node->_M_prev;
	return __tmp;
      }

      bool
      operator==(const _Self& __x) const _GLIBCXX_NOEXCEPT
      { return _M_node == __x._M_node; }

      bool
      operator!=(const _Self& __x) const _GLIBCXX_NOEXCEPT
      { return _M_node != __x._M_node; }

      // The only member points to the %list element.
      __detail::_List_node_base* _M_node;
    };
```